crates/accelerate/src/gate_direction.rs

```diff
@@ -0,0 +1,159 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::nlayout::PhysicalQubit;
+use crate::target_transpiler::{Qargs, Target};
+use pyo3::prelude::*;
+use pyo3::types::{PySet, PyTuple};
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{OperationRef, PyInstruction};
+use qiskit_circuit::{
+    dag_circuit::{DAGCircuit, NodeType},
+    error::DAGCircuitError,
+    operations::Operation,
+    packed_instruction::PackedInstruction,
+    Qubit,
+};
+
+/// Check if the two-qubit gates follow the right direction with respect to the coupling map.
+///
+/// Args:
+///     dag: the DAGCircuit to analyze
+///
+///     coupling_edges: set of edge pairs representing a directed coupling map, against which gate directionality is checked
+///
+/// Returns:
+///     true iff all two-qubit gates comply with the coupling constraints
+#[pyfunction]
+#[pyo3(name = "check_gate_direction_coupling")]
+fn py_check_with_coupling_map(
+    py: Python,
+    dag: &DAGCircuit,
+    coupling_edges: &Bound<PySet>,
```

Comment:
Good point. I switched to using `HashSet<(Qubit, Qubit)>`. This required adding `FromPyObject` implementation for `Qubit` which I'd appreciate a review of. 
I didn't end up using `PhysicalQubit` after all since `DAGCircuit` doesn't use it and I wasn't clear about whether or not it's safe or effective to force-cast the `DAGCircuit` indices in pass implementation. Let me know if there is an elegant way for that.